.github/workflows/release.yml

```diff
@@ -105,4 +105,458 @@ jobs:
+            -a "workflowref=${{ github.workflow_ref }}" \
+            -a "workflowsha=${{ github.workflow_sha }}" \
+            -a "type=paralus" \ 
+            ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:${{ steps.meta.outputs.tags }}
+
+          cosign sign --key cosign.key \
+            -a "tags"="${{ steps.meta-init.outputs.tags }}" \
+            -a "labels=${{ steps.meta-init.outputs.labels }}" \
+            -a "owner"=${{ github.repository_owner }}" \
+            -a "repo=${{ github.repository }}" \
+            -a "actor=${{ github.actor }}" \
+            -a "trigerringactor=${{ github.triggering_actor}}" \
+            -a "ref=${{ github.ref }}" \
+            -a "sha=${{ github.sha }}" \
+            -a "workflowref=${{ github.workflow_ref }}" \
+            -a "workflowsha=${{ github.workflow_sha }}" \
+            -a "type=paralus-init" \ 
+            ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}-init:${{ steps.meta-init.outputs.tags }}
+
+          cosign sign --key cosign.key \
+            -a "tags"="${{ steps.meta-sync.outputs.tags }}" \
+            -a "labels=${{ steps.meta-sync.outputs.labels }}" \
+            -a "owner"=${{ github.repository_owner }}" \
+            -a "repo=${{ github.repository }}" \
+            -a "actor=${{ github.actor }}" \
+            -a "trigerringactor=${{ github.triggering_actor}}" \
+            -a "ref=${{ github.ref }}" \
+            -a "sha=${{ github.sha }}" \
+            -a "workflowref=${{ github.workflow_ref }}" \
+            -a "workflowsha=${{ github.workflow_sha }}" \
+            -a "type=kratos-synchronizer" \ 
+            ${{ env.REGISTRY }}/paralusio/kratos-synchronizer:${{ steps.meta-sync.outputs.tags }}
+
+      # Removing private key from the local runner
+      # Sensitive information from local environments can be leaked.
+      # Using Key Management Services is recommended.
+      - name: Clean up the Cosign private key
+        run: |
+          rm cosign.key
+
+      - name: Write public key to disk
+        shell: bash
+        run: |
+            'echo "$KEY" > cosign.pub'
+        env:
+          KEY: ${{ secrets.COSIGN_PUBLIC_KEY }}
+      
+      # Using external action for simplicity sake.
+      # There codebase is available for public on GitHub thus vulnerabilities can be discovered if any.
+      - name: The Docker CLI
+        uses: actions-hub/docker@v1.0.3
+        env:
+          DOCKER_USERNAME: ${{ secrets.REGISTRY_USERNAME }}
+          DOCKER_PASSWORD: ${{ secrets.REGISTRY_PASSWORD }}
+          
+      - name: Verify Cosign Signature (paralus)
+        run: |
+
+         # Fetch image digest
+          image_digest=$(docker inspect --format='{{index .RepoDigests 0}}' ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:${{ steps.meta.outputs.tags }} | cut -d '@' -f 2)
+
+          # Verify the signature
+          verification_result=$(cosign verify --key cosign.pub ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:${{ steps.meta.outputs.tags }} | jq -r '.')
```

Comment:
I think by default cosign verify run couple of verification methods such as:

  - The cosign claims were validated
  - The signatures were verified against the specified public key
  - Any certificates were verified against the roots

Just wondering what does comparing verification result with data would mean and will it really necessary at all. Let us know if you've thought on this.

Reply:
Cosign verify only shows that the key itself hasn't been changed since it's creation. Now the annotation checks are required to verify who created the keys and from what source. I will explain it further in the above query where I have explained why I have used these annotations.